designs/disruption-controls-drain-timeout.md

```diff
@@ -0,0 +1,78 @@
+
+## Proposed Spec
+
+```yaml
+apiVersion: karpenter.sh/v1beta1
+kind: NodePool
+metadata:
+  name: default
+spec: # This is not a complete NodePool Spec.
+  disruption:
+    consolidationPolicy: WhenUnderutilized || WhenEmpty
+    consolidateAfter: 10m || Never # metav1.Duration
+    expireAfter: 10m || Never # Equivalent to v1alpha5 TTLSecondsUntilExpired
+    forceDrainAfter: 24h || Never
+```
+
+## Code Definition
+
+```go
+type Disruption struct {
+    {...}
+    // ForceDrainAfter is a nillable duration, parsed as a metav1.Duration.
+    // Users can use "Never" to disable force draining of pods.
+    ForceDrainAfter *NillableDuration `json:"forceDrainAfter" hash:"ignore"`
+}
+```
+
+## Validation/Defaults
+The `forceDrainAfter` fields accepts a common duration string which defaults to a value of `Never`. Omitting the field results in the default value instructing the controller to wait indefinitely for pods to drain gracefully, maintaining the existing behavior.
+
+## Prior Art
+This is the field from CAPI's MachineDeployment spec which implements identical behavior.
+```
+$ kubectl explain machinedeployment.spec.template.spec.nodeDrainTimeout
+KIND:     MachineDeployment
+VERSION:  cluster.x-k8s.io/v1beta1
+
+FIELD:    nodeDrainTimeout <string>
+
+DESCRIPTION:
+     NodeDrainTimeout is the total amount of time that the controller will spend
+     on draining a node. The default value is 0, meaning that the node can be
+     drained without any time limitations. NOTE: NodeDrainTimeout is different
+     from `kubectl drain --timeout`
+```
+
+## Implementation
+
+### Force Drain Detection
+If forceDrainAfter is set, I think it should be straightforward to have the node termination controller add a timestamp as an annotation on the Node resource denoting when the Node will be force drained at a specific time (is RFC-3339 the preferred standard?)
+
+When a drain fails and would be requeued to be attempted again, if the current time is after the timestamp of the forceDrainAt annotation, a requeue should be skipped and the force drain logic should be triggered instead.
+
+### Force Drain Behavior
+Drain behavior should be similar to using kubectl to bypass standard eviction protections.
+```
+$ kubectl drain my-node --grace-period=0 --disable-eviction=true
+
+$ kubectl drain --help
+Options:
+    --grace-period=-1:
+	Period of time in seconds given to each pod to terminate gracefully. If negative, the default value specified
+	in the pod will be used.
+
+    --disable-eviction=false:
+	Force drain to use delete, even if eviction is supported. This will bypass checking PodDisruptionBudgets, use
+	with caution.
+```
+
+It might sufficient to skip straight to termination of the node at the cloud provider rather than waiting for a successful force drain to complete.
```

Comment:
The problem with this is similar to https://github.com/kubernetes-sigs/karpenter/issues/655. You'll get leaked pods, which is a hefty consideration if we decide to go this way.

Reply:
Daemonsets tolerating all taints I see as a separate issue. Many daemonsets do this because it's the easiest way to ensure that the pods are running as long as possible, without having to think about all the different addons that may add taints. Whichever ones terminate all taints will either need to be okay with being leaked or we may need to respect their terminationGracePeriodSeconds as well.